mne_connectivity/decoding/decomposition.py

```diff
@@ -0,0 +1,474 @@
+                    "fmax": self.fmax,
+                    "bandwidth": self.mt_bandwidth,
+                    "adaptive": self.mt_adaptive,
+                    "low_bias": self.mt_low_bias,
+                }
+            )
+            csd = csd_array_multitaper(**csd_kwargs)
+        elif self.mode == "fourier":
+            csd_kwargs.update({"fmin": self.fmin, "fmax": self.fmax})
+            csd = csd_array_fourier(**csd_kwargs)
+        else:
+            csd_kwargs.update(
+                {
+                    "frequencies": np.arange(
+                        self.fmin,
+                        self.fmax + self.cwt_freq_resolution,
+                        self.cwt_freq_resolution,
+                    ),
+                    "n_cycles": self.cwt_n_cycles,
+                }
+            )
+            csd = csd_array_morlet(**csd_kwargs)
+
+        csd = csd.sum(self.fmin, self.fmax).get_data(index=0)
+        csd = np.reshape(csd, csd.shape[0] ** 2)
+
+        return np.expand_dims(csd, 1)
+
+    def _extract_filters_and_patterns(self):
+        """Extract filters and patterns from the connectivity estimator."""
+        # XXX: need to sort indices and transpose patterns when multiple comps supported
+        self.filters_ = (
+            self._conn_estimator.filters[0, 0, : len(self.indices[0]), 0],
+            self._conn_estimator.filters[1, 0, : len(self.indices[1]), 0],
+        )
+
+        self.patterns_ = (
+            self._conn_estimator.patterns[0, 0, : len(self.indices[0]), 0],
+            self._conn_estimator.patterns[1, 0, : len(self.indices[1]), 0],
+        )
+
+        # XXX: remove once support for multiple comps implemented
+        self.filters_ = (
+            np.expand_dims(self.filters_[0], 1),
+            np.expand_dims(self.filters_[1], 1),
+        )
+        self.patterns_ = (
+            np.expand_dims(self.patterns_[0], 0),
+            np.expand_dims(self.patterns_[1], 0),
+        )
+
+    def transform(self, X):
+        """Decompose data into connectivity sources using the fitted filters.
+
+        Parameters
+        ----------
+        X : array, shape=((n_epochs, ) n_signals, n_times)
+            The data to be transformed by the connectivity decomposition filters.
+
+        Returns
+        -------
+        X_transformed : array, shape=((n_epochs, ) n_components*2, n_times)
+            The transformed data. The first ``n_components`` channels are the
+            transformed seeds, and the last ``n_components`` channels are the
+            transformed targets.
+        """
+        self._check_X(X, ndim=(2, 3))
+        if self.filters_ is None:
+            raise RuntimeError(
+                "no filters are available, please call the `fit` method first"
+            )
+
+        # transform seed and target data
+        X_seeds = self.filters_[0].T @ X[..., self.indices[0], :]
+        X_targets = self.filters_[1].T @ X[..., self.indices[1], :]
```

Comment:
This assumes the decomposition is always linear and without a bias term. Technically the ABC class cannot know how the given `_conn_estimator` wants to perform the decomposition and intuitively I would think this should call `_conn_estimator.transform()`. Would something like that make sense in the larger design?

Reply:
I get what you mean that this transformation should be abstracted more. But in that case my first thought would be to handle this transformation in something like the `_AbstractCoherenceDecomposition` you describe above. I.e. a similar approach to how you suggest abstracting `fit` could be handled.

I just feel adding a `transform` method to the connectivity estimators would be overburdening these classes. Currently everything we are using them for (e.g. CSD rank subspace projection, filter and pattern computations, transforming filters and patterns back into channel space) is also essential for their role in the `spectral_connectivity_epochs/time` functions, but the transformation of new data would very niche to this use-case.

Reply:
> I just feel adding a `transform` method to the connectivity estimators would be overburdening these classes.

I don't feel that at all (or maybe I misunderstand the situation). All scikit-learn estimators have both `fit` and `transform`.

Reply:
Okay, maybe we can discuss this further in tomorrow's meeting.